In [90]:
# Import pandas, numpy and matplotlib libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns
from matplotlib.backends.backend_pdf import PdfPages
%matplotlib inline

# Convert csv into dataframes. Note that original categorical data has space in front of the value, this is
# eliminated with an extra parameter.
df = pd.read_csv('OnlineNewsPopularity.csv',  keep_default_na=True, sep=',\s+', delimiter=',', skipinitialspace=True)

# Make sure all rows/columns are visible
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# (1). [50] Prepare a data quality report for the CSV file.

### - Check how many rows and columns your CSV has.

In [91]:
#Print (rows, columns)
df.shape

(39644, 61)

We have a 39644 articles in our dataframe, with 33 features associated with each individual article.

### - Print the first and the last 5 rows.

In [92]:
df.head(5)

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  \
0                  0.815385        4.0             2.0       1.0         0.0   
1                  0.791946        3.0             1.0       1.0         0.0   
2                  0.663866        3.0             1.0       1.0         0.0   
3                  0.665635        9.0             0.0       1.0         0.0   
4                  0.540890       19.0            19.0      20.0         0.0   

   average_token_length  num_keywords  data_channel_is_lifestyle  \
0              4.680365           5.0                        0.0   
1              4.913725           4.0                        0.0   
2              4.393365           6.0                        0.0   
3              4.404896           7.0                        0.0   
4              4.682836           7.0                        0.0   

   data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  \
0                            1.0                  0.0                     0.0   
1                            0.0                  1.0                     0.0   
2                            0.0                  1.0                     0.0   
3                            1.0                  0.0                     0.0   
4                            0.0                  0.0                     0.0   

   data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  \
0                   0.0                    0.0         0.0         0.0   
1                   0.0                    0.0         0.0         0.0   
2                   0.0                    0.0         0.0         0.0   
3                   0.0                    0.0         0.0         0.0   
4                   1.0                    0.0         0.0         0.0   

   kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  \
0         0.0         0.0         0.0         0.0         0.0         0.0   
1         0.0         0.0         0.0         0.0         0.0         0.0   
2         0.0         0.0         0.0         0.0         0.0         0.0   
3         0.0         0.0         0.0         0.0         0.0         0.0   
4         0.0         0.0         0.0         0.0         0.0         0.0   

   kw_avg_avg  self_reference_min_shares  self_reference_max_shares  \
0         0.0                      496.0                      496.0   
1         0.0                        0.0                        0.0   
2         0.0                      918.0                      918.0   
3         0.0                        0.0                        0.0   
4         0.0                      545.0                    16000.0   

   self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  \
0                  496.000000                1.0                 0.0   
1                    0.000000                1.0                 0.0   
2                  918.000000                1.0                 0.0   
3                    0.000000                1.0                 0.0   
4                 3151.157895                1.0                 0.0   

   weekday_is_wednesday  weekday_is_thursday  weekday_is

In [93]:
df.tail(5)

url  timedelta  \
39639  http://mashable.com/2014/12/27/samsung-app-aut...        8.0   
39640  http://mashable.com/2014/12/27/seth-rogen-jame...        8.0   
39641  http://mashable.com/2014/12/27/son-pays-off-mo...        8.0   
39642     http://mashable.com/2014/12/27/ukraine-blasts/        8.0   
39643  http://mashable.com/2014/12/27/youtube-channel...        8.0   

       n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
39639            11.0             346.0         0.529052               1.0   
39640            12.0             328.0         0.696296               1.0   
39641            10.0             442.0         0.516355               1.0   
39642             6.0             682.0         0.539493               1.0   
39643            10.0             157.0         0.701987               1.0   

       n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  \
39639                  0.684783        9.0             7.0       1.0   
39640                  0.885057        9.0             7.0       3.0   
39641                  0.644128       24.0             1.0      12.0   
39642                  0.692661       10.0             1.0       1.0   
39643                  0.846154        1.0             1.0       0.0   

       num_videos  average_token_length  num_keywords  \
39639         1.0              4.523121           8.0   
39640        48.0              4.405488           7.0   
39641         1.0              5.076923           8.0   
39642         0.0              4.975073           5.0   
39643         2.0              4.471338           4.0   

       data_channel_is_lifestyle  data_channel_is_entertainment  \
39639                        0.0                            0.0   
39640                        0.0                            0.0   
39641                        0.0                            0.0   
39642                        0.0                            0.0   
39643                        0.0                            1.0   

       data_channel_is_bus  data_channel_is_socmed  data_channel_is_tech  \
39639                  0.0                     0.0                   1.0   
39640                  0.0                     1.0                   0.0   
39641                  0.0                     0.0                   0.0   
39642                  0.0                     0.0                   0.0   
39643                  0.0                     0.0                   0.0   

       data_channel_is_world  kw_min_min  kw_max_min  kw_avg_min  kw_min_max  \
39639                    0.0        -1.0       671.0     173.125     26900.0   
39640                    0.0        -1.0       616.0     184.000      6500.0   
39641                    0.0        -1.0       691.0     168.250      6200.0   
39642                    1.0        -1.0         0.0      -1.000         0.0   
39643                    0.0        -1.0        97.0      23.500    205600.0   

       kw_max_max     kw_avg_max   kw_min_avg   kw_max_avg   kw_avg_avg  \
39639    843300.0  374962.500000  2514.742857  4004.342857  3031.115764   
39640    843300.0  192985.714286  1664.267857  5470.168651  3411.660830   
39641    843300.0  295850.000000  1753.882353  6880.687034  4206.439195   
39642    843300.0  254600.000000     0.000000  3384.316871  1777.895883   
39643    843300.0  366200.000000  3035.080555  3613.512953  3296.909481   

       self_reference_min_shares  self_reference_max_shares  \
39639                    11400.0                    48000.0   
39640                     2100.0                     2100.0   
39641                     1400.0                     1400.0   
39642                      452.0                      452.0   
39643                     2100.0                     2100.0   

       self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  \
39639                37033.333333                0.0                 0.0   
39640                 2100.000000                0.0      

### - Convert the features to their appropriate data types (e.g., decide which features are more appropriate as continuous and which ones as categorical types). 

In [94]:
df.dtypes

url                               object
timedelta                        float64
n_tokens_title                   float64
n_tokens_content                 float64
n_unique_tokens                  float64
n_non_stop_words                 float64
n_non_stop_unique_tokens         float64
num_hrefs                        float64
num_self_hrefs                   float64
num_imgs                         float64
num_videos                       float64
average_token_length             float64
num_keywords                     float64
data_channel_is_lifestyle        float64
data_channel_is_entertainment    float64
data_channel_is_bus              float64
data_channel_is_socmed           float64
data_channel_is_tech             float64
data_channel_is_world            float64
kw_min_min                       float64
kw_max_min                       float64
kw_avg_min                       float64
kw_min_max                       float64
kw_max_max                       float64
kw_avg_max      

We have concluded that it might be more efficient to convert the features relating to data channel type (eg data_channel_is_lifestyle) and days of the week (eg, weekday_is_monday) on which an article was published into categorical features.
This would mean a single categorical feature 'weekday', with seven levels (ie monday through to sunday).
And another categorical feature 'data_channel', which would have six levels (lifestyle, entertainment, socmed, bus, tech, world)

In [95]:
df.head()

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  \
0                  0.815385        4.0             2.0       1.0         0.0   
1                  0.791946        3.0             1.0       1.0         0.0   
2                  0.663866        3.0             1.0       1.0         0.0   
3                  0.665635        9.0             0.0       1.0         0.0   
4                  0.540890       19.0            19.0      20.0         0.0   

   average_token_length  num_keywords  data_channel_is_lifestyle  \
0              4.680365           5.0                        0.0   
1              4.913725           4.0                        0.0   
2              4.393365           6.0                        0.0   
3              4.404896           7.0                        0.0   
4              4.682836           7.0                        0.0   

   data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  \
0                            1.0                  0.0                     0.0   
1                            0.0                  1.0                     0.0   
2                            0.0                  1.0                     0.0   
3                            1.0                  0.0                     0.0   
4                            0.0                  0.0                     0.0   

   data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  \
0                   0.0                    0.0         0.0         0.0   
1                   0.0                    0.0         0.0         0.0   
2                   0.0                    0.0         0.0         0.0   
3                   0.0                    0.0         0.0         0.0   
4                   1.0                    0.0         0.0         0.0   

   kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  \
0         0.0         0.0         0.0         0.0         0.0         0.0   
1         0.0         0.0         0.0         0.0         0.0         0.0   
2         0.0         0.0         0.0         0.0         0.0         0.0   
3         0.0         0.0         0.0         0.0         0.0         0.0   
4         0.0         0.0         0.0         0.0         0.0         0.0   

   kw_avg_avg  self_reference_min_shares  self_reference_max_shares  \
0         0.0                      496.0                      496.0   
1         0.0                        0.0                        0.0   
2         0.0                      918.0                      918.0   
3         0.0                        0.0                        0.0   
4         0.0                      545.0                    16000.0   

   self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  \
0                  496.000000                1.0                 0.0   
1                    0.000000                1.0                 0.0   
2                  918.000000                1.0                 0.0   
3                    0.000000                1.0                 0.0   
4                 3151.157895                1.0                 0.0   

   weekday_is_wednesday  weekday_is_thursday  weekday_is

Insert new column 'weekday', with a placeholder value in it for the moment.

In [96]:
df['weekday'] = 'no day yet'
df.head()
        

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  \
0                  0.815385        4.0             2.0       1.0         0.0   
1                  0.791946        3.0             1.0       1.0         0.0   
2                  0.663866        3.0             1.0       1.0         0.0   
3                  0.665635        9.0             0.0       1.0         0.0   
4                  0.540890       19.0            19.0      20.0         0.0   

   average_token_length  num_keywords  data_channel_is_lifestyle  \
0              4.680365           5.0                        0.0   
1              4.913725           4.0                        0.0   
2              4.393365           6.0                        0.0   
3              4.404896           7.0                        0.0   
4              4.682836           7.0                        0.0   

   data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  \
0                            1.0                  0.0                     0.0   
1                            0.0                  1.0                     0.0   
2                            0.0                  1.0                     0.0   
3                            1.0                  0.0                     0.0   
4                            0.0                  0.0                     0.0   

   data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  \
0                   0.0                    0.0         0.0         0.0   
1                   0.0                    0.0         0.0         0.0   
2                   0.0                    0.0         0.0         0.0   
3                   0.0                    0.0         0.0         0.0   
4                   1.0                    0.0         0.0         0.0   

   kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  \
0         0.0         0.0         0.0         0.0         0.0         0.0   
1         0.0         0.0         0.0         0.0         0.0         0.0   
2         0.0         0.0         0.0         0.0         0.0         0.0   
3         0.0         0.0         0.0         0.0         0.0         0.0   
4         0.0         0.0         0.0         0.0         0.0         0.0   

   kw_avg_avg  self_reference_min_shares  self_reference_max_shares  \
0         0.0                      496.0                      496.0   
1         0.0                        0.0                        0.0   
2         0.0                      918.0                      918.0   
3         0.0                        0.0                        0.0   
4         0.0                      545.0                    16000.0   

   self_reference_avg_sharess  weekday_is_monday  weekday_is_tuesday  \
0                  496.000000                1.0                 0.0   
1                    0.000000                1.0                 0.0   
2                  918.000000                1.0                 0.0   
3                    0.000000                1.0                 0.0   
4                 3151.157895                1.0                 0.0   

   weekday_is_wednesday  weekday_is_thursday  weekday_is

Change value of new 'weekeday' column to appropriate value based on the value of the other 'days of the week' columns

In [97]:
for i, row in df.iterrows():
    if df.at[i,'weekday_is_monday']==1.0:
        df.at[i, 'weekday'] = 'monday'
    elif df.at[i,'weekday_is_tuesday']==1.0:
        df.at[i, 'weekday'] = 'tuesday'
    elif df.at[i, 'weekday_is_wednesday']==1.0:
        df.at[i, 'weekday'] = 'wednesday'
    elif df.at[i, 'weekday_is_thursday']==1.0:
        df.at[i, 'weekday'] = 'thursday'
    elif df.at[i,'weekday_is_friday']==1.0:
        df.at[i, 'weekday'] = 'friday'
    elif df.at[i, 'weekday_is_saturday']==1.0:
        df.at[i, 'weekday'] = 'saturday'
    elif df.at[i,'weekday_is_sunday']==1.0:
        df.at[i, 'weekday'] = 'sunday'

Test that the values have been assigned correctly to the column 'weekday' by temporarily creating a new dataframe, so the weekday columns can all be viewed together

In [98]:
newdf = df[['weekday', 'weekday_is_monday','weekday_is_tuesday', 'weekday_is_wednesday', 'weekday_is_thursday','weekday_is_friday', 'weekday_is_saturday', 'weekday_is_sunday']].copy()

In [99]:
newdf.head()

weekday  weekday_is_monday  weekday_is_tuesday  weekday_is_wednesday  \
0  monday                1.0                 0.0                   0.0   
1  monday                1.0                 0.0                   0.0   
2  monday                1.0                 0.0                   0.0   
3  monday                1.0                 0.0                   0.0   
4  monday                1.0                 0.0                   0.0   

   weekday_is_thursday  weekday_is_friday  weekday_is_saturday  \
0                  0.0                0.0                  0.0   
1                  0.0                0.0                  0.0   
2                  0.0                0.0                  0.0   
3                  0.0                0.0                  0.0   
4                  0.0                0.0                  0.0   

   weekday_is_sunday  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0

Drop the columns 'weekday_is_monday', 'weekday_is_tuesday', 'weekday_is_wednesday', 'weekday_is_thursday', 'weekday_is_friday', 'weekday_is_saturday', 'weekday_is_sunday' from the dataframe, as we no longer need them after creating the 'weekday' column. 

In [100]:
df = df.drop('weekday_is_monday', 1)
df = df.drop('weekday_is_tuesday', 1)
df = df.drop('weekday_is_wednesday', 1)
df = df.drop('weekday_is_thursday', 1)
df = df.drop('weekday_is_friday', 1)
df = df.drop('weekday_is_saturday', 1)
df = df.drop('weekday_is_sunday', 1)

In [101]:
df.head()

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  \
0                  0.815385        4.0             2.0       1.0         0.0   
1                  0.791946        3.0             1.0       1.0         0.0   
2                  0.663866        3.0             1.0       1.0         0.0   
3                  0.665635        9.0             0.0       1.0         0.0   
4                  0.540890       19.0            19.0      20.0         0.0   

   average_token_length  num_keywords  data_channel_is_lifestyle  \
0              4.680365           5.0                        0.0   
1              4.913725           4.0                        0.0   
2              4.393365           6.0                        0.0   
3              4.404896           7.0                        0.0   
4              4.682836           7.0                        0.0   

   data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  \
0                            1.0                  0.0                     0.0   
1                            0.0                  1.0                     0.0   
2                            0.0                  1.0                     0.0   
3                            1.0                  0.0                     0.0   
4                            0.0                  0.0                     0.0   

   data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  \
0                   0.0                    0.0         0.0         0.0   
1                   0.0                    0.0         0.0         0.0   
2                   0.0                    0.0         0.0         0.0   
3                   0.0                    0.0         0.0         0.0   
4                   1.0                    0.0         0.0         0.0   

   kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  \
0         0.0         0.0         0.0         0.0         0.0         0.0   
1         0.0         0.0         0.0         0.0         0.0         0.0   
2         0.0         0.0         0.0         0.0         0.0         0.0   
3         0.0         0.0         0.0         0.0         0.0         0.0   
4         0.0         0.0         0.0         0.0         0.0         0.0   

   kw_avg_avg  self_reference_min_shares  self_reference_max_shares  \
0         0.0                      496.0                      496.0   
1         0.0                        0.0                        0.0   
2         0.0                      918.0                      918.0   
3         0.0                        0.0                        0.0   
4         0.0                      545.0                    16000.0   

   self_reference_avg_sharess  is_weekend    LDA_00    LDA_01    LDA_02  \
0                  496.000000         0.0  0.500331  0.378279  0.040005   
1                    0.000000         0.0  0.799756  0.050047  0.050096   
2                  918.000000         0.0  0.217792  0.033334  0.033351   
3                    0.000000         0.0  0.028573  0.419300  0.494651   
4                 3151.157895         0.0  0.028633  0.028794  0.028575   

     LDA_03    LDA_04  global_subjecti

Repeat process for 'data_channel_is_xyz' columns - ie, create new column called 'data_channel' and populate it with values based on the values of the existing channels 'data_channel_is_lifestyle', 'data_channel_is_entertainment', etc.

In [102]:
# create new column and insert placeholder value
df['data_channel'] = 'no channel yet'
df.head(1000)

url  timedelta  \
0    http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1    http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2    http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3    http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4     http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   
5     http://mashable.com/2013/01/07/beewi-smart-toys/      731.0   
6    http://mashable.com/2013/01/07/bodymedia-armba...      731.0   
7     http://mashable.com/2013/01/07/canon-poweshot-n/      731.0   
8    http://mashable.com/2013/01/07/car-of-the-futu...      731.0   
9    http://mashable.com/2013/01/07/chuck-hagel-web...      731.0   
10   http://mashable.com/2013/01/07/cosmic-events-d...      731.0   
11    http://mashable.com/2013/01/07/crayon-creatures/      731.0   
12       http://mashable.com/2013/01/07/creature-cups/      731.0   
13           http://mashable.com/2013/01/07/dad-jokes/      731.0   
14   http://mashable.com/2013/01/07/downton-abbey-t...      731.0   
15   http://mashable.com/2013/01/07/earth-size-plan...      731.0   
16           http://mashable.com/2013/01/07/echo-game/      731.0   
17   http://mashable.com/2013/01/07/entrepreneur-tr...      731.0   
18   http://mashable.com/2013/01/07/facebook-sick-app/      731.0   
19   http://mashable.com/2013/01/07/felt-audio-puls...      731.0   
20        http://mashable.com/2013/01/07/ford-glympse/      731.0   
21    http://mashable.com/2013/01/07/ftc-google-leaks/      731.0   
22   http://mashable.com/2013/01/07/fujifilm-50x-su...      731.0   
23   http://mashable.com/2013/01/07/hillary-clinton...      731.0   
24              http://mashable.com/2013/01/07/htc-q1/      731.0   
25   http://mashable.com/2013/01/07/huawei-ascend-m...      731.0   
26   http://mashable.com/2013/01/07/iheartradio-app...      731.0   
27   http://mashable.com/2013/01/07/intel-awesome-l...      731.0   
28   http://mashable.com/2013/01/07/isp02-iphone-ta...      731.0   
29      http://mashable.com/2013/01/07/jobs-contently/      731.0   
30   http://mashable.com/2013/01/07/land-a-job-at-s...      731.0   
31   http://mashable.com/2013/01/07/lego-taking-rob...      731.0   
32    http://mashable.com/2013/01/07/lg-oled-tv-price/      731.0   
33   http://mashable.com/2013/01/07/lionel-messi-fa...      731.0   
34   http://mashable.com/2013/01/07/living-room-pin...      731.0   
35   http://mashable.com/2013/01/07/livio-connect-e...      731.0   
36         http://mashable.com/2013/01/07/magisto-ces/      731.0   
37   http://mashable.com/2013/01/07/man-instagram-i...      731.0   
38       http://mashable.com/2013/01/07/mcafee-spying/      731.0   
39   http://mashable.com/2013/01/07/messi-year-all-...      731.0   
40   http://mashable.com/2013/01/07/monster-ea-gami...      731.0   
41   http://mashable.com/2013/01/07/monster-katana-...      731.0   
42   http://mashable.com/2013/01/07/nectar-mobile-p...      731.0   
43   http://mashable.com/2013/01/07/nvidia-project-...      731.0   
44      http://mashable.com/2013/01/07/nvidia-tegra-4/      731.0   
45   http://mashable.com/2013/01/07/obama-boombox-i...      731.0   
46   http://mashable.com/2013/01/07/omron-fitness-m...      731.0   
47   http://mashable.com/2013/01/07/one-directions-...      731.0   
48   http://mashable.com/2013/01/07/panasonic-secon...      731.0   
49   http://mashable.com/2013/01/07/pantech-discove...      731.0   
50            http://mashable.com/2013/01/07/papertab/      731.0   
51   http://mashable.com/2013/01/07/philips-fidelio...      731.0   
52   http://mashable.com/2013/01/07/philips-fidelio...      731.0   
53   http://mashable.com/2013/01/07/pitchfork-advan...      731.0   
54   http://mashable.com/2013/01/07/polaroid-androi...      731.0   
55   http://mashable.com/2013/01/07/power-matters-a...      731.0   
56   http://mashable.com/2013/01/07/rage-comics-dying/      731.0   
57      http://mashable.com/2013/01/07/reeddit-re

In [103]:
for i, row in df.iterrows():
    if df.at[i,'data_channel_is_lifestyle']==1.0:
        df.at[i, 'data_channel'] = 'lifestyle'
    elif df.at[i,'data_channel_is_entertainment']==1.0:
        df.at[i, 'data_channel'] = 'entertainment'
    elif df.at[i, 'data_channel_is_socmed']==1.0:
        df.at[i, 'data_channel'] = 'socmed'
    elif df.at[i, 'data_channel_is_tech']==1.0:
        df.at[i, 'data_channel'] = 'tech'
    elif df.at[i,'data_channel_is_bus']==1.0:
        df.at[i, 'data_channel'] = 'bus'
    elif df.at[i, 'data_channel_is_world']==1.0:
        df.at[i, 'data_channel'] = 'world'

In [104]:
df.head()

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  num_videos  \
0                  0.815385        4.0             2.0       1.0         0.0   
1                  0.791946        3.0             1.0       1.0         0.0   
2                  0.663866        3.0             1.0       1.0         0.0   
3                  0.665635        9.0             0.0       1.0         0.0   
4                  0.540890       19.0            19.0      20.0         0.0   

   average_token_length  num_keywords  data_channel_is_lifestyle  \
0              4.680365           5.0                        0.0   
1              4.913725           4.0                        0.0   
2              4.393365           6.0                        0.0   
3              4.404896           7.0                        0.0   
4              4.682836           7.0                        0.0   

   data_channel_is_entertainment  data_channel_is_bus  data_channel_is_socmed  \
0                            1.0                  0.0                     0.0   
1                            0.0                  1.0                     0.0   
2                            0.0                  1.0                     0.0   
3                            1.0                  0.0                     0.0   
4                            0.0                  0.0                     0.0   

   data_channel_is_tech  data_channel_is_world  kw_min_min  kw_max_min  \
0                   0.0                    0.0         0.0         0.0   
1                   0.0                    0.0         0.0         0.0   
2                   0.0                    0.0         0.0         0.0   
3                   0.0                    0.0         0.0         0.0   
4                   1.0                    0.0         0.0         0.0   

   kw_avg_min  kw_min_max  kw_max_max  kw_avg_max  kw_min_avg  kw_max_avg  \
0         0.0         0.0         0.0         0.0         0.0         0.0   
1         0.0         0.0         0.0         0.0         0.0         0.0   
2         0.0         0.0         0.0         0.0         0.0         0.0   
3         0.0         0.0         0.0         0.0         0.0         0.0   
4         0.0         0.0         0.0         0.0         0.0         0.0   

   kw_avg_avg  self_reference_min_shares  self_reference_max_shares  \
0         0.0                      496.0                      496.0   
1         0.0                        0.0                        0.0   
2         0.0                      918.0                      918.0   
3         0.0                        0.0                        0.0   
4         0.0                      545.0                    16000.0   

   self_reference_avg_sharess  is_weekend    LDA_00    LDA_01    LDA_02  \
0                  496.000000         0.0  0.500331  0.378279  0.040005   
1                    0.000000         0.0  0.799756  0.050047  0.050096   
2                  918.000000         0.0  0.217792  0.033334  0.033351   
3                    0.000000         0.0  0.028573  0.419300  0.494651   
4                 3151.157895         0.0  0.028633  0.028794  0.028575   

     LDA_03    LDA_04  global_subjecti

In order to check that the correct values have been assigned to the new columns, create a new df so we can see just the relevant columns side by side

In [105]:
newdf = df[['data_channel', 'data_channel_is_world','data_channel_is_bus', 'data_channel_is_tech', 'data_channel_is_socmed','data_channel_is_entertainment', 'data_channel_is_lifestyle']].copy()

In [106]:
newdf.head()

data_channel  data_channel_is_world  data_channel_is_bus  \
0  entertainment                    0.0                  0.0   
1            bus                    0.0                  1.0   
2            bus                    0.0                  1.0   
3  entertainment                    0.0                  0.0   
4           tech                    0.0                  0.0   

   data_channel_is_tech  data_channel_is_socmed  \
0                   0.0                     0.0   
1                   0.0                     0.0   
2                   0.0                     0.0   
3                   0.0                     0.0   
4                   1.0                     0.0   

   data_channel_is_entertainment  data_channel_is_lifestyle  
0                            1.0                        0.0  
1                            0.0                        0.0  
2                            0.0                        0.0  
3                            1.0                        0.0  
4                            0.0                        0.0

Now we can drop the columns 'data_channel_is_world' etc as we have a 'data_channel' column to replace these six columns.

In [107]:
df = df.drop('data_channel_is_world', 1)
df = df.drop('data_channel_is_bus', 1)
df = df.drop('data_channel_is_tech', 1)
df = df.drop('data_channel_is_socmed', 1)
df = df.drop('data_channel_is_entertainment', 1)
df = df.drop('data_channel_is_lifestyle', 1)

In [108]:
df.tail()

url  timedelta  \
39639  http://mashable.com/2014/12/27/samsung-app-aut...        8.0   
39640  http://mashable.com/2014/12/27/seth-rogen-jame...        8.0   
39641  http://mashable.com/2014/12/27/son-pays-off-mo...        8.0   
39642     http://mashable.com/2014/12/27/ukraine-blasts/        8.0   
39643  http://mashable.com/2014/12/27/youtube-channel...        8.0   

       n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
39639            11.0             346.0         0.529052               1.0   
39640            12.0             328.0         0.696296               1.0   
39641            10.0             442.0         0.516355               1.0   
39642             6.0             682.0         0.539493               1.0   
39643            10.0             157.0         0.701987               1.0   

       n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  \
39639                  0.684783        9.0             7.0       1.0   
39640                  0.885057        9.0             7.0       3.0   
39641                  0.644128       24.0             1.0      12.0   
39642                  0.692661       10.0             1.0       1.0   
39643                  0.846154        1.0             1.0       0.0   

       num_videos  average_token_length  num_keywords  kw_min_min  kw_max_min  \
39639         1.0              4.523121           8.0        -1.0       671.0   
39640        48.0              4.405488           7.0        -1.0       616.0   
39641         1.0              5.076923           8.0        -1.0       691.0   
39642         0.0              4.975073           5.0        -1.0         0.0   
39643         2.0              4.471338           4.0        -1.0        97.0   

       kw_avg_min  kw_min_max  kw_max_max     kw_avg_max   kw_min_avg  \
39639     173.125     26900.0    843300.0  374962.500000  2514.742857   
39640     184.000      6500.0    843300.0  192985.714286  1664.267857   
39641     168.250      6200.0    843300.0  295850.000000  1753.882353   
39642      -1.000         0.0    843300.0  254600.000000     0.000000   
39643      23.500    205600.0    843300.0  366200.000000  3035.080555   

        kw_max_avg   kw_avg_avg  self_reference_min_shares  \
39639  4004.342857  3031.115764                    11400.0   
39640  5470.168651  3411.660830                     2100.0   
39641  6880.687034  4206.439195                     1400.0   
39642  3384.316871  1777.895883                      452.0   
39643  3613.512953  3296.909481                     2100.0   

       self_reference_max_shares  self_reference_avg_sharess  is_weekend  \
39639                    48000.0                37033.333333         0.0   
39640                     2100.0                 2100.000000         0.0   
39641                     1400.0                 1400.000000         0.0   
39642                      452.0                  452.000000         0.0   
39643                     2100.0                 2100.000000         0.0   

         LDA_00    LDA_01    LDA_02    LDA_03    LDA_04  global_subjectivity  \
39639  0.025038  0.025001  0.151701  0.025000  0.773260             0.482679   
39640  0.029349  0.028575  0.231866  0.681635  0.028575             0.564374   
39641  0.159004  0.025025  0.025207  0.643794  0.146970             0.510296   
39642  0.040004  0.040003  0.839987  0.040002  0.040004             0.358578   
39643  0.050001  0.799339  0.050000  0.050659  0.050001             0.517893   

       global_sentiment_polarity  global_rate_positive_words  \
39639                   0.141964                    0.037572   
39640                   0.194249                    0.039634   
39641                   0.024609                    0.033937   
39642                  -0.008066                    0.020528   
39643                   0.104892                    0.063694   

       global_rate_negative_words  rate_positive_words  rate_negative_words  \
39639                  

Check new columns are of the correct datatypes - we want them to be categorical.

In [109]:
df.dtypes

url                              object
timedelta                       float64
n_tokens_title                  float64
n_tokens_content                float64
n_unique_tokens                 float64
n_non_stop_words                float64
n_non_stop_unique_tokens        float64
num_hrefs                       float64
num_self_hrefs                  float64
num_imgs                        float64
num_videos                      float64
average_token_length            float64
num_keywords                    float64
kw_min_min                      float64
kw_max_min                      float64
kw_avg_min                      float64
kw_min_max                      float64
kw_max_max                      float64
kw_avg_max                      float64
kw_min_avg                      float64
kw_max_avg                      float64
kw_avg_avg                      float64
self_reference_min_shares       float64
self_reference_max_shares       float64
self_reference_avg_sharess      float64


In [110]:
# Select columns containing categorical data
categorical_columns = df[['weekday', 'data_channel']].columns
#income was chosen as category because the data contains single digit values representing income categories

# Convert data type to category for these columns
for column in categorical_columns:
    df[column] = df[column].astype('category')

df.dtypes

url                               object
timedelta                        float64
n_tokens_title                   float64
n_tokens_content                 float64
n_unique_tokens                  float64
n_non_stop_words                 float64
n_non_stop_unique_tokens         float64
num_hrefs                        float64
num_self_hrefs                   float64
num_imgs                         float64
num_videos                       float64
average_token_length             float64
num_keywords                     float64
kw_min_min                       float64
kw_max_min                       float64
kw_avg_min                       float64
kw_min_max                       float64
kw_max_max                       float64
kw_avg_max                       float64
kw_min_avg                       float64
kw_max_avg                       float64
kw_avg_avg                       float64
self_reference_min_shares        float64
self_reference_max_shares        float64
self_reference_a

In [111]:
# Print table with descriptive statistics for all the categorical features
df.select_dtypes(['category']).describe().T

count unique        top  freq
weekday       39644      7  wednesday  7435
data_channel  39644      7      world  8427

#### NB NOTE TO TEAM
NB we can see there are some missing values from data_channel (there should only be six unique values, as there are six channels in total, but there are seven unique values listed here - some of these are still the placeholder value 'no channel yet', where no channel was assigned. We will have to decide what to do about these!

### - Drop duplicate rows and columns, if any.

In [112]:
# check for duplicate rows
print('Duplicate rows:', df.duplicated()[df.duplicated() == True].shape[0])

Duplicate rows: 0


In [113]:
# Check for duplicate columns
print('Duplicate columns:',df.columns.size - df.columns.unique().size)

Duplicate columns: 0
